### DISCUSSION TITLE
Unable to find user object

### DISCUSSION THREAD

**Discussion by rtaylo45** [idaholab/moose#18513] [Q&A General] [answered]
Hello,

Im trying to build and use a user object i defined using the action system but im running into this error when trying to access the object after i build it, `Unable to find user object with name spec`. I made a previous post about trying to have a user object add variables and was told that i should build the user object and adding the variables in an custom action block. See here https://github.com/idaholab/moose/discussions/18483. Here is the code im using in my `AddSpecies` action. 

```
#include "AddSpeciesAction.h"
#include "SpeciesDefinition.h"
#include "FEProblemBase.h"
#include "AddVariableAction.h"

registerMooseAction("moleApp", AddSpeciesAction, "add_variable");
registerMooseAction("moleApp", AddSpeciesAction, "add_species_definition");

InputParameters
AddSpeciesAction::validParams()
{
  InputParameters params = Action::validParams();
  params.addRequiredParam<std::vector<std::string>>("species_names", "The species names");
  params.addParam<std::vector<std::string>>("phase_names", "Phase names for the species");
  params.addRequiredParam<std::vector<Real>>("molar_masses", "The molar masses of species");
  params.addClassDescription("Adds the species to the problem");
  return params;
}

AddSpeciesAction::AddSpeciesAction(const InputParameters & params)
  : Action(params)
{
}

void
AddSpeciesAction::act()
{
  if (_current_task == "add_species_definition")
  {
    auto params = _factory.getValidParams("SpeciesDefinition");
    params.set<std::vector<std::string>>("species_names") =
      getParam<std::vector<std::string>>("species_names");
    params.set<std::vector<Real>>("molar_masses") =
      getParam<std::vector<Real>>("molar_masses");
    if (isParamValid("phase_names"))
      params.set<std::vector<std::string>>("phase_names") =
        getParam<std::vector<std::string>>("phase_names");
    _problem->addUserObject("SpeciesDefinition", "spec", params);
  }
  else if (_current_task == "add_variable")
  {
    auto _vars = _problem->getUserObject<SpeciesDefinition>
      ("spec").getSpeciesVarNames();
    auto fe_type = AddVariableAction::feType(_pars);
    auto type = AddVariableAction::determineType(fe_type, 1);
    auto var_params = _factory.getValidParams(type);

    var_params.applySpecificParameters(_pars, {"family", "order", "scaling"});

    for (auto & var : _vars)
      _problem->addVariable(type, var, var_params);
  }
}
```
I added the following to my mainApp.C
```
  registerSyntax("ConvectionAction", "Convection");
  registerSyntax("AddSpeciesAction", "AddSpecies");
  registerMooseObjectTask("add_species_definition", AddSpeciesAction, false);
  addTaskDependency("add_species_definition", "add_user_object");
```
Im not sure if there is some kinda order to when `add_species_definition` is being called after `add_variable` or something. Also, if i call `_problem->getUserObject<SpeciesDefinition>("spec")` in other actions in my app will it return the correct user object that is created in this action?

**Comment by dschwen** [comment]
Change 
```
registerMooseObjectTask("add_species_definition", AddSpeciesAction, false);
````

to

```
registerTask("add_species_definition", false);
```

**Reply by rtaylo45** [reply]
So this did not fix my issue. It still gives me the same error. On a side note, Im not exactly sure what line `addTaskDependency("add_species_definition", "add_user_object");` does but if i comment this line out, the code runs and gives me a different error. It seg falts on line `_problem->addUserObject("SpeciesDefinition", "spec", params);`

**Reply by dschwen** [reply]
Hmm. Can you share your repo?

**Reply by rtaylo45** [reply]
Unfortunately it is hosted on an internal ORNL git repo so I cannot. I can post my main apps C file and the header and C file for the action object, if that would help.

**Comment by cpgr** [comment, marked as answer]
I think that your problem is that `add_user_object` (and hence `add_species_definition`) gets run after `add_variables`, see https://github.com/idaholab/moose/blob/next/framework/src/base/Moose.C. 

`addTaskDependency("add_species_definition", "add_user_object")` does what you probably thought it does - it makes sure that the `add_user_object` task is run before your `add_species_definition` task runs. So you definitely want that line left in!

You could possibly just include the add variable code into the `add_species_definition` to fix it, or even make another task like `add_species_variables` to add your variables and make it depend on your `add_species_definition` task.

**Reply by rtaylo45** [reply]
This work! Thanks for your help. This also helped me understand what the `addTaskDependency` call means.

**Reply by dschwen** [reply]
Adding variables in or after the `add_user_object` task seems a little dangerous, as by that time the `setup_variables_complete` task, the setup of the quadrature rules (dependent on variable order), and the setup of periodic BCs have already run and wouldn't know anything about the new variables.

Can you infer the list of variable names to set up with out actually instantiating the user object? E.g. using a static member method?

**Reply by rtaylo45** [reply]
Oh ok i see what you are talking about. Yes i can infer the names, ill try to change the order of the calls making adding the variables dependent on `add_variable`

**Reply by cpgr** [reply]
@dschwen - good point!
